orderbook/src/api/order_validation.rs

```diff
@@ -0,0 +1,734 @@
+    pub buy_token_balance: BuyTokenDestination,
+    pub sell_token_balance: SellTokenSource,
+}
+
+impl From<Order> for PreOrderData {
+    fn from(order: Order) -> Self {
+        Self {
+            owner: order.order_meta_data.owner,
+            sell_token: order.order_creation.sell_token,
+            buy_token: order.order_creation.buy_token,
+            receiver: order.actual_receiver(),
+            valid_to: order.order_creation.valid_to,
+            buy_token_balance: order.order_creation.buy_token_balance,
+            sell_token_balance: order.order_creation.sell_token_balance,
+        }
+    }
+}
+
+impl OrderValidator {
+    pub fn new(
+        code_fetcher: Box<dyn CodeFetching>,
+        native_token: WETH9,
+        banned_users: Vec<H160>,
+        min_order_validity_period: Duration,
+        fee_validator: Arc<dyn MinFeeCalculating>,
+        bad_token_detector: Arc<dyn BadTokenDetecting>,
+        balance_fetcher: Arc<dyn BalanceFetching>,
+    ) -> Self {
+        Self {
+            code_fetcher,
+            native_token,
+            banned_users,
+            min_order_validity_period,
+            fee_validator,
+            bad_token_detector,
+            balance_fetcher,
+        }
+    }
+
+    async fn pre_validate(&self, order: PreOrderData) -> Result<(), PreValidationError> {
+        if self.banned_users.contains(&order.owner) {
+            return Err(PreValidationError::Forbidden);
+        }
+        if order.buy_token_balance != BuyTokenDestination::Erc20 {
+            return Err(PreValidationError::UnsupportedBuyTokenDestination(
+                order.buy_token_balance,
+            ));
+        }
+        if !matches!(
+            order.sell_token_balance,
+            SellTokenSource::Erc20 | SellTokenSource::External
+        ) {
+            return Err(PreValidationError::UnsupportedSellTokenSource(
+                order.sell_token_balance,
+            ));
+        }
+        if order.valid_to
+            < shared::time::now_in_epoch_seconds() + self.min_order_validity_period.as_secs() as u32
+        {
+            return Err(PreValidationError::InsufficientValidTo);
+        }
+        if has_same_buy_and_sell_token(&order, &self.native_token) {
+            return Err(PreValidationError::SameBuyAndSellToken);
+        }
+        if order.buy_token == BUY_ETH_ADDRESS {
+            let code_size = self
+                .code_fetcher
+                .code_size(order.receiver)
+                .await
+                .map_err(PreValidationError::Other)?;
+            if code_size != 0 {
+                return Err(PreValidationError::TransferEthToContract);
+            }
+        }
+        Ok(())
+    }
+
+    pub async fn post_validate(
+        &self,
+        order: Order,
+        sender: Option<H160>,
+    ) -> Result<(), ValidationError> {
+        self.pre_validate(PreOrderData::from(order.clone()))
+            .await
+            .map_err(ValidationError::Pre)?;
```

Comment:
I find it weird that `post_validate` calls `pre_validate`. From the names I would expect you to have to call both:
```
validator.pre_validate()?;
// do stuff...
validator.post_validate()?;
```

Maybe `pre-`/`post-` prefixes can be changed (`partial-`/`full-`? Maybe something else)?

Reply:
Yes, initially I had a call to each independently, but the only place this is done would be in `Orderbook.add_order`. I was worried that future refactors might forget that then both need to be called, so I enforced this by calling the pre inside the post. I completely agree that the name could be changed to `partial` and `full` or maybe even `partial_validate` and just `validate`.

Reply:
Fun fact, I had already mentioned this here https://github.com/gnosis/gp-v2-services/pull/1176#discussion_r720812626

Reply:
> `maybe even partial_validate and just validate.`

Works for me. @vkgnosis, from your comments, I think you also didn't find the names super clear, do you have any suggestions as well?